### PR TITLE
store/tikv: avoid unnecessary pessimistic rollback on ErrKeyExists

### PR DESCRIPTION
**Pull request by sticnarf**
### What problem does this PR solve?

This PR fixes https://github.com/pingcap/tidb/issues/23799 on release-5.0.

### What is changed and how it works?

This PR corrects the type matching of `ErrKeyExists` and adds corresponding tests.

### Check List 

Tests 

- Unit test

### Release note 

- Avoid unnecessary pessimistic rollback when pessimistic locking receives ErrKeyExists

### CODE REVIEW

File: store/tikv/txn.go [hunk 1]
```diff
@@ -469,7 +469,8 @@ func (txn *KVTxn) LockKeys(ctx context.Context, lockCtx *kv.LockCtx, keysInput .
 					txn.us.UnmarkPresumeKeyNotExists(key)
 				}
 			}
-			keyMayBeLocked := terror.ErrorNotEqual(kv.ErrWriteConflict, err) && terror.ErrorNotEqual(kv.ErrKeyExists, err)
+			_, isErrKeyExist := errors.Cause(err).(*ErrKeyExist)
+			keyMayBeLocked := terror.ErrorNotEqual(kv.ErrWriteConflict, err) && !isErrKeyExist
```
Comment (youjiali1995) [inline store/tikv/txn.go:473]: Why do we need to use different ways here? The deadlock error below also uses `dl, ok := errors.Cause(err).(*ErrDeadlock);`

Comment (sticnarf) [reply store/tikv/txn.go:473]: Do you mean embedding the downcast in the `if` expression like
```
if _, isErrKeyExist := errors.Cause(err).(*ErrKeyExist); len(keys) > 1 || (terror.ErrorNotEqual(kv.ErrWriteConflict, err) && !isErrKeyExist) {
```

I'm afraid it looks too long...

Comment (sticnarf) [reply store/tikv/txn.go]: #23686 fixed the same bug on master with a helper function. So I change this PR to be consistent with the master.

**Review by youjiali1995 — APPROVED** [review]

**Review by AndreMouche — APPROVED** [review]
LGTM

**Pull request merged**